File: LINKED LIST PROGRAMS/Multi-Level-Doubly-LinkedList/multiLevelDoublyLinkedList.c
#include<stdio.h>
#include<stdlib.h>

typedef struct _Node
{
    struct _Node* next;
    struct _Node* pre;
    struct _Node* nextLevel;
    int data;
}Node;

Node* createNode(int data)
{
    Node* t = (Node*)malloc(sizeof(Node));
    t->next = NULL;
    t->nextLevel = NULL;
    t->pre = NULL;
    t->data = data;
    return t;
}

Node* generateMultiLevelDoublyLinkedListFromArray(int* arr, int size)
{
    if(size == 0) return NULL;
    Node* start = createNode(arr[0]);
    Node* currLevelStart = start;
    Node *t = start;
    int i;
    Node* curr;
    i = 1;
    while(i<size)
    {
        int data = arr[i];
        if(data == 0)
        {
            Node* nextLevelStart = currLevelStart;
            i++;
            //10 20 30 0 0 0 0 40 50
            while(i<size)
            {
                data = arr[i];
                if(data == 0)
                {
                    if(nextLevelStart->next) nextLevelStart = nextLevelStart->next;
                }
                else
                {
                    break;
                }
                i++;
            }
            if(i>=size)return start;
            nextLevelStart->nextLevel = createNode(data);
            // printf("\nlevel started with %d after %d ,",data, nextLevelStart->data);
            t = nextLevelStart->nextLevel;
            currLevelStart = t;
            i++;
        }
        else
        {
            // printf("added %d,", data);
            curr = createNode(data);
            curr->pre = t;
            t->next = curr;
            t = curr;
            i++;
        }
    }
    return start;
}

void flattern(Node* start)
{
    Node* t = start;
    while(t!=NULL)
    {
        printf("%d ", t->data);
        if(t->nextLevel) flattern(t->nextLevel);
        t = t->next;
    }
}

int main()
{
    int arr[] = {10,20,30,40,50,0,0,60,70,80,0,90,100};
    // 10 20 30 40 50
    //    60 70 80
    //    90 100

    // for(int i=0;i<sizeof(arr)/sizeof(arr[0]);i++) printf("%d ", arr[i]);
    Node* start = generateMultiLevelDoublyLinkedListFromArray(arr, sizeof(arr)/sizeof(arr[0]));
    flattern(start);
    printf("\n");
    return 0;
}

File: LINKED LIST PROGRAMS/Multi-Level-Doubly-LinkedList/multiLevelDoublyLinkedList.java
import java.util.*;

class Node
{
    int data;
    Node next, prev, nextLevel, prevLevel;
    Node(int data){
        this.data = data;
        this.next = this.prev = this.nextLevel = this.prevLevel = null;
    }
    public String toString(){
        return ""+data;
    }
}

class DLL
{
    Node start;
    DLL(){start = null;}
    void generateMultiLevelDoublyLinkedListFromArray(int arr[]){
        int size = arr.length;
        if(size == 0) return;
        start = new Node(arr[0]);
        Node t = start;
        Node currLevelStart = start;
        int i = 1;
        while(i < size){
            int data = arr[i];
            if(data == 0)
            {
                Node nextLevelStart = currLevelStart;
                i++;
                while(i<size){
                    data = arr[i];
                    if(data != 0) break;
                    if(nextLevelStart.next != null) nextLevelStart = nextLevelStart.next;
                    i++;
                }
                if(i == size) break;
                nextLevelStart.nextLevel = new Node(data);
                nextLevelStart.nextLevel.prevLevel = nextLevelStart;
                currLevelStart = nextLevelStart.nextLevel;
                t = currLevelStart;
            }
            else
            {
                Node curr = new Node(data);
                t.next = curr;
                curr.prev = t;
                t = curr;
            }
            i++;
        }
    }
    void flattern(Node node){
        if(node==null) return;
        System.out.print(node+" ");
        if(node.nextLevel!=null){
            flattern(node.nextLevel);
        }
        flattern(node.next);
    }
    void flattern(){flattern(this.start);}
}

class multiLevelDoublyLinkedList
{
    public static void main(String[] args) 
    {
        int arr[] = {10,20,30,40,50,0,0,60,70,80,0,90,100};
        for(int i : arr) System.out.print(i + " ");
        System.out.println();
        DLL list = new DLL();
        list.generateMultiLevelDoublyLinkedListFromArray(arr);
        list.flattern();
    }
}